wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -510,44 +623,208 @@ private void ensurePreDestroyIsNotExecuted() throws ServerException {
             this.runtime = runtime;
         }
 
         private RuntimeDescriptor(RuntimeDescriptor descriptor) {
             this(new WorkspaceRuntimeImpl(descriptor.runtime));
-            this.isStopping = descriptor.isStopping;
+            this.status = descriptor.status;
         }
 
-        /**
-         * Returns an {@link Optional} describing a started {@link WorkspaceRuntime},
-         * if the runtime is in starting state then an empty {@code Optional} will be returned.
-         */
+        /** Returns the instance of {@code WorkspaceRuntime} described by this descriptor. */
         public WorkspaceRuntimeImpl getRuntime() {
             return runtime;
         }
 
         /**
-         * Returns the status of the started workspace runtime.
-         * The relation between {@link #getRuntime()} and this method
-         * is pretty clear, whether workspace is in starting state, then
-         * {@code getRuntime()} will return an empty optional, otherwise
-         * the optional describing a running or stopping workspace runtime.
+         * Returns the status of the {@code WorkspaceRuntime} described by this descriptor.
+         * Never returns {@link WorkspaceStatus#STOPPED} status, you'll rather get {@link NotFoundException}
+         * from {@link #get(String)} method.
          */
         public WorkspaceStatus getRuntimeStatus() {
-            if (isStopping) {
-                return WorkspaceStatus.STOPPING;
+            return status;
+        }
+
+        private void setRuntimeStatus(WorkspaceStatus status) {
+            this.status = status;
+        }
+    }
+
+    @VisibleForTesting
+    class AddMachineEventSubscriber implements EventSubscriber<MachineStatusEvent> {
+        @Override
+        public void onEvent(MachineStatusEvent event) {
+            if (event.getEventType() == MachineStatusEvent.EventType.RUNNING) {
+                try {
+                    final MachineImpl machine = machineManager.getMachine(event.getMachineId());
+                    if (!addMachine(machine)) {
+                        machineManager.destroy(machine.getId(), true);
+                    }
+                } catch (NotFoundException | MachineException x) {
+                    LOG.warn(format("An error occurred during an attempt to add the machine '%s' to the workspace '%s'",
+                                    event.getMachineId(),
+                                    event.getWorkspaceId()),
+                             x);
+                }
             }
-            if (runtime.getDevMachine() == null) {
-                return WorkspaceStatus.STARTING;
+        }
+    }
+
+    /**
+     * Tries to add the machine to the WorkspaceRuntime.
+     *
+     * @return true if machine is added or will be added to the corresponding WorkspaceRuntime,
+     * returns false when incoming machine can't be added and should be destroyed.
+     */
+    @VisibleForTesting
```

Comment:
and I still don't get why anyone couldn't subclasses and override this existing code. With the "keep everything private" it's tending to avoid something pluggable, overridable and extensible.